### PR TITLE
GeoJSON functions

### PR DESCRIPTION
**Pull request by ajfriend**

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -14,10 +14,7 @@ jobs:
     strategy:
       matrix:
         os: [macos-latest, windows-latest, ubuntu-20.04]
-        python-version: ['3.5', '3.6', '3.7', '3.8', '3.9', '3.10', '3.11']
-        exclude:
-          - os: windows-latest
-            python-version: 3.5
+        python-version: ['3.6', '3.7', '3.8', '3.9', '3.10', '3.11']
```
Comment (dfellis) [inline .github/workflows/tests.yml:17]: Python 3.12 released recently, btw. But it may also be worth doing that in a different PR because they finally killed some deprecated features of `setup.py` and I don't know if we're impacted by any of them.

Comment (dfellis) [reply .github/workflows/tests.yml:17]: (So, maybe add `'3.12'` to this PR, but if it fails, roll it back and we dig into it in a follow-up PR?)

Comment (ajfriend) [reply .github/workflows/tests.yml:17]: I'm anticipating some work to figure out the changes with 3.12, which will also involve updating the wheels matrix. I'd rather just do all the 3.12 work together in a separate PR.

File: src/h3/_h3shape.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+Helpers for cells_to_geojson and geojson_to_cells.
+
+Dealing with GeoJSON Polygons and MultiPolygons can be confusing because
+there are so many nested lists. To help keep track, we use the following
+symbols to denote different levels of nesting.
+
+LL0: lat/lng or lng/lat pair
+LL1: list of LL0s
+LL2: list of LL1s (i.e., a polygon with holes)
+LL3: list of LL2s (i.e., several polygons with holes)
```
Comment (jongbinjung) [inline src/h3/_h3shape.py:171]: 👍

File: src/h3/api/_api_template.py [hunk 3]
```diff
@@ -426,43 +427,78 @@ def polygon_to_cells(self, polygon, res):
...
         """
-        Return a list of h3.Polygon objects describing the area
-        covered by a set of H3 cells.
+        Return a H3MultiPoly describing the area covered by a set of H3 cells.
 
         Parameters
         ----------
         cells : iterable of H3 cells
+        tight : bool
+            If True, return H3Poly if possible. If False, always return H3MultiPoly
```
Comment (jongbinjung) [inline src/h3/api/_api_template.py:451]: Is this helpful?

I can see how this can be convenient for many cases, but maybe confusing/more work to deal with for the caller if the return type may be one of two things, and it's not easy to know in advance what they're getting.

File: src/h3/api/_api_template.py [hunk 4]
```diff
@@ -426,43 +427,78 @@ def polygon_to_cells(self, polygon, res):
...
+        out = H3MultiPoly(*polys)
+
+        if tight and len(out) == 1:
+            out = out[0]
+
+        return out
+
+    def geo_to_cells(self, geo, res):
+        """ Convert from __geo_interface__ to cells.
+        Parameters
```
Comment (jongbinjung) [inline src/h3/api/_api_template.py:479]:
```suggestion
        """ Convert from __geo_interface__ to cells.
        
        Parameters
```

nit

File: src/h3/_h3shape.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+
+
+class H3Shape(metaclass=ABCMeta):
+    @property
+    @abstractmethod
+    def __geo_interface__(self):
+        """ https://github.com/pytest-dev/pytest-cov/issues/428 """
+
+
+class H3Poly(H3Shape):
```
[the remark this answers is not in the document]
Comment (jongbinjung) [reply src/h3/_h3shape.py:11]: I like this. I can totally see myself getting confused otherwise ....

[the remark this answers is not in the document]
Comment (ajfriend) [reply src/h3/_h3shape.py:11]: Just to make sure we're talking about the same design goals, I think there are two potential points of confusion we could try to address:

1. Make it clear that these objects are different from `shapely.Polygon` and `shapely.MultiPolygon`, which has confused some users in the past.
2. Additionally, make it clear that this polygon type accepts coordinates in lat/lng order.

I was originally just trying to address point 1, and figured folks could look at docstrings to learn the lat/lng input format. But you both think there is value in making it super explicit in the name itself?

If we do that, then the `H3` prefix isn't needed anymore (the object would already be clearly distinct from `shapely` objects), so we could drop it to make it a bit shorter: `LatLngPoly` and `LatLngMultiPoly`?

[the remark this answers is not in the document]
Comment (nrabinowitz) [reply src/h3/_h3shape.py:11]: That works for me. I just think the lat/lng vs lng/lat order has probably been the single biggest source of user questions and bugs, and clarifying that seems important.

File: src/h3/api/_api_template.py (hunk 3 above)
Comment (ajfriend) [reply src/h3/api/_api_template.py:451]: It is easy to know in advance what they're getting if they set `tight=False` :)

I could go either way on this, but I found it ergonomic when I'm playing with data sets manually, or if I'm demonstrating how things work with smaller examples.

And it seems to be standard practice, at least with `geopandas`, to use the smaller container when possible.
For example:

```
>>> geopandas.read_file(geodatasets.get_path('geoda airbnb')).geometry
0     POLYGON ((-87.60914 41.84469, -87.60915 41.844...
1     POLYGON ((-87.59215 41.81693, -87.59231 41.816...
2     POLYGON ((-87.62880 41.80189, -87.62879 41.801...
3     POLYGON ((-87.60671 41.81681, -87.60670 41.816...
4     POLYGON ((-87.59215 41.81693, -87.59215 41.816...
                            ...                        
72    POLYGON ((-87.69646 41.70714, -87.69644 41.706...
73    POLYGON ((-87.64215 41.68508, -87.64249 41.685...
74    MULTIPOLYGON (((-87.83658 41.98640, -87.83658 ...
75    POLYGON ((-87.65456 41.99817, -87.65456 41.998...
76    POLYGON ((-87.80676 42.00084, -87.80676 42.000...

File: src/h3/_h3shape.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ajfriend) [reply src/h3/_h3shape.py:11]: Alternatively, is it worth considering switching to lng/lat order for `h3-py` to match `geopandas` and GeoJSON Would that reduce the volume of user questions and bugs? Or is it not worth the mismatch with the other bindings?

[the remark this answers is not in the document]
Comment (nrabinowitz) [reply src/h3/_h3shape.py:11]: Seems like that might undermine future efforts to more efficiently ship the structure to C?

[the remark this answers is not in the document]
Comment (nrabinowitz) [reply src/h3/_h3shape.py:11]: Either way, I'd include the order in the name

[the remark this answers is not in the document]
Comment (ajfriend) [reply src/h3/_h3shape.py:11]: > Seems like that might undermine future efforts to more efficiently ship the structure to C?

True. And, in that case, we'd probably have to change `h3.latlng_to_cell()`. We can keep it as is.

File: src/h3/_h3shape.py [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    ...      (37.73, -122.43)],
+    ... )
+    <H3Poly: [4/(3, 5)]>
+    """
+    def __init__(self, outer, *holes):
+        loops = [outer] + list(holes)
+        for loop in loops:
+            if len(loop) in (1, 2):
+                self.outer = []
+                self.holes = []
```
[the remark this answers is not in the document]
Comment (ajfriend) [reply src/h3/_h3shape.py:56]: Fair. I'll remove
